Be more consistent in live_var_analysis

miscco opened the pull request:
The naming convention changes within the same function. 

Stay consistent and use the temporary flag inserted

Btw is there any reason this is a std::set rather than an std::unordered_set?

I dont really see the necessity of the ordering

miscco commented:
I should say that i havent had time to fully test this

daniel-schuermann commented:
Thanks! Could you change the target to daniel-schuermann:backend and rebase?
The master branch is currently only for bugfixes.
IIRC, not using unordered_set was just me being lazy to think how to best provide a hash function. It also really depends on the problem size when unordered_set becomes faster than set. I didn't really test, but many blocks are quite small, or for them a logarithmic access is preferable. But I'm open to any suggestions.

miscco commented:
Ok, although it might take some time as I need to upgrade my linux environement on a "when the kids sleep" basis.

I found that in `live_var_analysis`the set is iterated from the end so there seems to be some ordering involved.

miscco commented:
So it does indeed build ;)

Which test suite to you run to test stuff? Although here it seems trivial

daniel-schuermann commented:
We are using pipelines from various games with vkpipeline-db, then VK-GL-CTS (although not for every patch as it takes quite long) and test games manually. Your patch seems to not affect any shader, so it's fine to be merged. Can you please rebase on top of backend?

miscco commented:
It should already be rebased

daniel-schuermann <notifications@github.com> schrieb am Fr., 5. Juli 2019
15:26:

> We are using pipelines from various games with vkpipeline-db, then
> VK-GL-CTS (although not for every patch as it takes quite long) and test
> games manually. Your patch seems to not affect any shader, so it's fine to
> be merged. Can you please rebase on top of backend?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/daniel-schuermann/mesa/pull/48?email_source=notifications&email_token=AAXATIGPSOEPDLVUT2NYNWTP55DXBA5CNFSM4H5ZY2JKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZJQSFI#issuecomment-508758293>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXATIHWU5ZQ2CRSDPYGVXTP55DXBANCNFSM4H5ZY2JA>
> .
>

miscco commented:
Interestingly one can now direcly switch the branch in the GitHub Gui.

daniel-schuermann commented:
No, it's not properly rebased. There should be only 1 commit in your MR, but you have some commits from the master branch like "Add DOOM app profile" which we don't want to push into the backend branch.

miscco commented:
I fixed some whitespace issues on the way and cleaned the branch up

daniel-schuermann commented:
Do you want to get a bit more involved and understand
1) what phi-instructions are and
2) how this pass calculates the register demand per instruction?
Then, I'd have a nice task for you to fix the PoE-crash. Let me know!

miscco commented:
I do know that phi instructions merge control flow in SSA forms and actually tried to understand the pass a bit better although there are some details that need some more considerations.

So yes I would like to have a look at PoE. If only to help a friend who likes it ;)

daniel-schuermann commented:
So then, here you go:
As hardware has no instruction to execute phis itself, we have to lower phis with copies. In ACO, this happens after register allocation in ssa_elimination(). For each phi, where definition register and operand register aren't the same, we insert a copy from the operand to the definition register at the corresponding predecessor.
Now the problem:
We distinguish between linear and (logical) phi. The exact meaning is not so important here. The copies for the linear phis are inserted immediately before the branch of the predecessor and for (logical) phis, we insert the copies right before aco_opcode::p_logical_end.
For calculating the register pressure, we don't consider this exactly. phi operands are just live-out of their predecessors (increasing the register pressure there).
But sometimes, there are (logical) phis with sgpr operands. And this means, they can increase the register pressure between p_logical_end and the end of the block although the copy (which kills them) is inserted before the p_logical_end. (I hope you can still follow me ;) )
The calculated register_demand is therefor too high between p_logical_end and the end of the block.
To fix this, I'd suppose we just remember the number of registers by which we have to decrease the register_demand in that area. To calculate this number, check if a phi operand is actually a kill (i.e. was not live before encountering the phi), the definition is vgpr and the operand is sgpr. The sum of the sizes of these operands is the number of registers which has to be decreased at the corresponding predecessor (the i-th phi-operand corresponds to the i-th logical predecessor).
Take your time to think about and feel free to ask. If you think that's too much theory or takes you too much time, don't worry. :)

Edit: just to clarify why the sgpr register demand matters in this case: in the broken PoE shaders, we reload some sgpr variables after p_logical_end, and thus, the compiler thinks we don't have enough registers there, but we do have enough because the phi operands are copied away before that.

daniel-schuermann commented:
I have to think about this issue a bit more. The idea doesn't completely resolve it.

daniel-schuermann commented:
I think, it's fine to do it this way. The related problem is register allocation which I'll take care of.

miscco commented:
Hm I just wanted to change a tiny thing and got taken away. 

Seems merge conflicts it is

daniel-schuermann commented:
yeah, well. we rebased this branch on top of upstream/master.
Easiest way to rebase your branch is probably:
git fetch upstream (if you called this repository upsteam)
git reset upstream/backend --hard
git cherry-pick your commits
or I just saw they became more, so you could use git format-patch to spill them into a file and load them with git am.

miscco commented:
I rebased them however it seems i need a way to find a newer libdrm-intel as the one i get from apt is too old ... grml

miscco commented on src/amd/compiler/aco_live_var_analysis.cpp:
```diff
@@ -77,7 +77,7 @@ void process_live_temps_per_block(Program *program, live& lives, Block* block, s
       /* substract the 2 sgprs from exec */
       if (exec_live)
          assert(sgpr_demand >= 2);
-      register_demand[idx] = {sgpr_demand - (exec_live ? 2 : 0), vgpr_demand};
```
I think I found a bug here, sgpr and vgpr are switched

miscco commented:
@d

daniel-schuermann commented on src/amd/compiler/aco_live_var_analysis.cpp:
```diff
@@ -70,6 +80,23 @@ void process_live_temps_per_block(Program *program, live& lives, Block* block, s
+                       std::inserter(live_sgprs, live_sgprs.end()),
+                       std::inserter(live_vgprs, live_vgprs.end()),
+                       [](Temp temp) { return temp.is_linear(); });
+
```
I that faster? I'm asking because it would be really cool to get this algorithm faster. It is used multiple times in the backend...
Please fix the rebase artifacts (I know they are removed in a later commit, but they shouldn't appear in the first place).

miscco replied:
Yeah that one I am not sure about. I think that it is in the realm of "not measurably slower"

What it brings to the table is a clear separation of concerns and a more declarative style. I dont think that it matters with small blocks though I would have to measure it.

daniel-schuermann commented on src/amd/compiler/aco_ir.h:
```diff
@@ -975,6 +975,11 @@ struct RegisterDemand {
       return *this;
    }
 
+   void update_pressure(const RegisterDemand other) {
```
maybe rename to just "update"?

daniel-schuermann commented on src/amd/compiler/aco_live_var_analysis.cpp:
```diff
@@ -244,8 +244,7 @@ live live_var_analysis(Program* program,
       process_live_temps_per_block(program, result, &program->blocks[block_idx], worklist);
-      new_demand.vgpr = std::max(new_demand.vgpr, program->blocks[block_idx].vgpr_demand);
-      new_demand.sgpr = std::max(new_demand.sgpr, program->blocks[block_idx].sgpr_demand);
+      new_demand.update_pressure(RegisterDemand(program->blocks[block_idx].vgpr_demand, program->blocks[block_idx].sgpr_demand));
```
RegisterDemand() is not explicit, so I'd say {vgpr_demand, sgpr_demand} is sufficient here.

miscco replied:
It will be cleanup up in a follow-up commit when `aco::Block` gets a RegisterDemand rather than the two `int16_t`

daniel-schuermann commented:
Wow, didn't see that coming. W.r.t p_logical_end thing, I already took care of it as it was quite urgent.
Unfortunately, this means you probably have to rebase again on that patch (will land in the next couple of hours, but it's only a minor change to live_var_analysis). Overall, the changes look good, but need a bit of cleanup.
E.g. you don't need the patches 'Cleanup handling of KILL/GEN' if you rework that anyway in the later commits. Also, if you introduce RegisterDemand, place it directly above Block instead of moving it in a later commit. And finally, I'd appreciate if you move the patches with the constexpr changes before any changes to live_var_analysis so that things stay together.
Thx!

miscco commented:
I have cleaned the PR up. I will rebase it later this evening when you have pushed your changes

miscco commented:
I have rebased on https://github.com/daniel-schuermann/mesa/commit/1f6010dd013342fce62a9f4fd55186cf83e4570a however due to the missing libdrm i cannot build it right now and I dont know howmuch time I have in the comming days.

I would be awesome if you could check it once.

miscco commented:
Man sometimes I simply hate git. rebased and fixed some merge conflicts...

daniel-schuermann commented:
I know what you mean ;)
Thx for rebasing. I'll try to go over it do do some tests, tomorrow.

miscco commented:
As said in the now gone commit I would finish this PR now and do the functional changes separately

miscco commented:
I dont know how it happens that the commits always get mixed up... I will have a look at my repo in the evening. I am pretty sure that I did all the squashing an merge-fixes at elast twice already

daniel-schuermann commented:
I feel almost sorry already, but maybe it's a good way to learn how to prepare proper PRs :)
Anyway, I felt so free to rebase your branch and squashe the commits together. I hope, that is okay from your side. If you disagree, just force-push your local branch again and my changes will be gone.

I build-tested your series, and got 3 warnings which you might want to fix:
../../../../../../Projects/mesa/src/amd/compiler/aco_live_var_analysis.cpp:213:62: warning: narrowing conversion of ‘((256 / ((int)program->aco::Program::num_waves)) & (~3))’ from ‘int’ to ‘int16_t’ {aka ‘short int’} inside { } [-Wnarrowing]
       program->max_reg_demand = { (256 / program->num_waves) & ~3, std::min<int16_t>(((total_sgpr_regs / program->num_waves) & ~7) - 2, max_addressible_sgpr)};
                                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~
[222/289] Compiling C++ object 'src/amd/compiler/3c53451@@aco@sta/aco_lower_to_cssa.cpp.o'.
../../../../../../Projects/mesa/src/amd/compiler/aco_lower_to_cssa.cpp: In function ‘void aco::{anonymous}::hoist_copies(aco::{anonymous}::cssa_ctx&)’:
../../../../../../Projects/mesa/src/amd/compiler/aco_lower_to_cssa.cpp:146:71: warning: comparison of integer expressions of different signedness: ‘unsigned int’ and ‘int16_t’ {aka ‘short int’} [-Wsign-compare]
                 reg_pressure.vgpr + info.phi->getDefinition(0).size() <= ctx.program->max_reg_demand.vgpr) {
                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../../../../../Projects/mesa/src/amd/compiler/aco_lower_to_cssa.cpp:152:71: warning: comparison of integer expressions of different signedness: ‘unsigned int’ and ‘int16_t’ {aka ‘short int’} [-Wsign-compare]
                 reg_pressure.sgpr + info.phi->getDefinition(0).size() <= ctx.program->max_reg_demand.sgpr) {

The rest of the nitpicks, I'll just comment directly in the commits.

daniel-schuermann commented on src/amd/compiler/aco_ir.h:
```diff
@@ -856,6 +856,76 @@ enum block_kind {
+      return *this;
+   }
+
+   RegisterDemand& operator-=(const Operand op) {
```
I don't think we need this function. If an Operand is not a temporary then it doesn't use a register before. Thus, wherever this function is used, we can also use -= op.getTemp()

miscco replied:
Yeah I was not sure about Constants and whether they can appear at that point, so I played save

daniel-schuermann commented on src/amd/compiler/aco_ir.h:
```diff
@@ -856,6 +856,76 @@ enum block_kind {
+
+   RegisterDemand operator+(const Temp t) {
+      if (t.type() == vgpr)
+         return RegisterDemand( vgpr + t.size(), sgpr );         
```
trailing whitespace

daniel-schuermann commented on src/amd/compiler/aco_live_var_analysis.cpp:
```diff
@@ -56,36 +53,29 @@ void process_live_temps_per_block(Program *program, live& lives, Block* block,
-            sgpr_demand += it->size();
-      }
+         new_demand += temp;
+      }           
```
trailing whitespace

daniel-schuermann commented on src/amd/compiler/aco_live_var_analysis.cpp:
```diff
@@ -209,31 +188,29 @@ void process_live_temps_per_block(Program *program, live& lives, Block* block,
    }
 
-   assert(block->index != 0 || (vgpr_demand == 0 && (sgpr_demand == 0)));
+   assert(block->index != 0 || (new_demand.vgpr == 0 && (new_demand.sgpr == 0)));
```
That we used unnecessary braces doesn't mean you are allowed to do this as well :P
maybe just assert(block->index != 0 || new_demand == RegisterDemand(0, 0))

daniel-schuermann commented on src/amd/compiler/aco_scheduler.cpp:
```diff
@@ -68,45 +68,35 @@ void move_element(T& list, size_t idx, size_t before) {
-         changes.second -= op.size();
-      else
-         changes.first -= op.size();
+      changes -= op;
```
I think that's the only use of the operand version. so, just use changes -= op.getTemp()

miscco replied:
Yeah, if it can never be a constant then we can skip this

miscco commented:
Hm I am 99% sure that on my local repo I have them already squashed. I wont be able to check that before the weekend though

miscco commented:
I removed the warnings and fixed the comments

miscco commented:
So I have fixed an temporary bug I introduced in between and finished aco_spill.

For the sake of not getting any more rebase errors I will stop working on this and the other PR so that we can finally review/merge this

daniel-schuermann commented:
I merged the first 3 commits as they seemed fine. The next commit on the list seems to contain a bug somewhere as register allocation asserts at (regs_free >= size) which probably means that the calculated register_demand is way too low (not updating correctly somewhere?). I force-pushed your branch so that only the latter 3 commits are on top now.

miscco commented:
Hm i went through the whole changelog and I cannot find a place where vgpr and sgpr are mixed or the wrong operator is found. 

Did you update the branch before merging? There was a intermittent bug in the now first patch where I switched sgpr and vgrp one. However that was fixed today

The one thing I could imagine is that first/second of a pair is switched somehow but the whole purpose of the change was to make that impossible.

miscco commented on src/amd/compiler/aco_ir.h:
```diff
@@ -856,6 +856,68 @@ enum block_kind {
+   }
+
+   RegisterDemand operator+(const RegisterDemand other) {
+      return RegisterDemand(vgpr + other.vgpr, sgpr + other.vgpr);
```
Argh

miscco commented on src/amd/compiler/aco_ir.h:
```diff
@@ -856,6 +856,68 @@ enum block_kind {
+   }
+
+   RegisterDemand operator-(const RegisterDemand other) {
+      return RegisterDemand(vgpr - other.vgpr, sgpr - other.vgpr);
```
And more :(

miscco commented on src/amd/compiler/aco_ir.h:
```diff
@@ -856,6 +856,68 @@ enum block_kind {
+
+   RegisterDemand& operator+=(const RegisterDemand other) {
+      vgpr += other.vgpr;
+      sgpr += other.vgpr;
```
I need a coffee

miscco replied:
or better a beer

miscco commented on src/amd/compiler/aco_ir.h:
```diff
@@ -856,6 +856,68 @@ enum block_kind {
+
+   RegisterDemand& operator-=(const RegisterDemand other) {
+      vgpr -= other.vgpr;
+      sgpr -= other.vgpr;
```
This is getting embarassing. Damn you autocorrecting brain

daniel-schuermann commented on src/amd/compiler/aco_ir.h:
```diff
@@ -856,6 +856,68 @@ enum block_kind {
+   }
+
+   RegisterDemand operator+(const Temp t) {
+      if (t.type() == vgpr)
```
Maybe we should have made RegType type-safe.
The problem is that you named the vgpr RegDemand exactly the same as RegType::vgpr, and the stupid compiler picks the wrong one ;)
Please change this (and the ones below) to
`if (t.type() == RegType::sgpr)`

It should check for sgpr instead of vgpr, because we also have the RegType::linear_vgpr which increases vgpr-demand. so, it's easier to only check for sgpr.

miscco replied:
Yikes, I will change that over the weekend ...

I need to get my test setup running

daniel-schuermann commented:
Thx! I tested all 3 patches against a large amount of shaders, and it seems that the second patch (Use RegisterDemand for sched_ctx) introduces a regression, although it shouldn't change anything.
I haven't look yet what goes wrong, though. The other 2 patches don't seem to affect anything, which is fine.

miscco commented:
Hm that is really strange. Could it be that there is an overflow in the determination of 
```
ctx.max_registers = { int16_t((256 / ctx.num_waves) & ~3), 
                      std::min<int16_t>(((total_sgpr_regs / ctx.num_waves) & ~7) - 2, max_addressible_sgpr)};
```

daniel-schuermann commented:
No, num_waves = [1,10].
Also, the change apparently only happens on shaders which execute aco_spill (this is only the case if the sgpr demand exceeds). I can try to find out what goes wrong on monday. If you want to check for yourself, I recommend to use https://gitlab.freedesktop.org/mesa/vkpipeline-db to grab some shaders and compare wh. affected games are e.g.
Talos Principle, the newer Tombraider games, F1 2017, warhammer 2, Evil Within, Doom.

Oh, I just saw this line:
new_demand.update(program->blocks[block_idx].register_demand) being called in live_var_analysis().
What happens if we call live_var_analysis twice? It must re-compute the register demand from the beginning.

miscco commented:
Hm that does not explain why things would break. As far as i can tell there is no functional change here

daniel-schuermann commented:
Sorry, I somehow messed up my tests. It's indeed the first commit which changes the output, the "Refactor handling of register demand".

daniel-schuermann commented on src/amd/compiler/aco_register_allocation.cpp:
```diff
@@ -812,7 +812,7 @@ bool get_reg_specified(ra_ctx& ctx,
    if (rc.type() == vgpr) {
       lb = 256;
-      ub = 256 + ctx.program->max_vgpr;
+      ub = 256 + ctx.program->max_reg_demand.sgpr;
```
found it! With that fixed, there are no changes to the output anymore.

miscco commented:
Thanks a lot i looked over the changes about 5 times and never catched it...
